## Bargheet 2023 (infant microbiota): ![[Bargheet2023_Fig_Example_Visualization.png]]
Något som det känns att vi ska göra, speciellt A, eventuellt även B och C.

 Metagenome profiling?
- **Determine the realtive abundance of microbial taxa**
	- High-quality reads were subjected to the CHOCOPhlAn  database using MetaPhlAn3 v.3.0.7 36 to determine the relative abundances of microbial taxa. The `merge metaphlan tables.py` script was used to merge relative abundance tables.
- **Assemble into longer contigs**
	- The quality-filtered short-read sequences were assembled into longer contiguous sequences (contigs) using metaSPAdes v.3.15.0 with the default parameters. For the assessment of assemblies performed by metaSPAdes, MetaQUAST from QUAST  v.5.0.2 was used with the -m 1000 option (Supplementary Table S1).
- **Predict ORF from assembled contigs using Prokka**
	- Open reading frames (ORFs) were predicted from the assembled contigs using Prokka v.1.14.5, with –kingdom Bacteria –centre X  –compliant –mincontig 200 options. Next, redundant ORFs were collapsed to one sequence using CD-HIT v.4.8.1, with the following parameters: -M 0 -T 0 -c 0.95 -n 8. The ORFs were annotated by searching nucleotide sequences against the Comprehensive Antibiotic Resistance Database (CARD v.3.0.9) 37 using ABRicate v.1.0.1 with –minid 80 –mincov 80 parameters. To estimate the abundance of the annotated ARGs, all fasta files (with nucleotide sequences) generated by Prokka were concatenated in one file.

Senare?
- Differentially abundant features between time points within treatment groups at family, genera, and species levels were identified using the DESeq2 v.1.34.0 package42 with the p values attained by the Wald test were corrected for multiple testing using the Benjamini Hochberg method.


## Pärnänen 2019 (milk microbiota)
Metagenomic analysis
- Metaphlan2 (Species level community profiling based on marker genes)
- Metaxa2 (Additional community profiling, 16S rRNA read extraction, paired-end mode)
